active/0003-extensible-resource-deployment-observability-improvements.md

```diff
@@ -0,0 +1,230 @@
+4. Conduct comprehensive testing in Canary regions to validate the functionality and stability of the API changes.
+5. Enable the feature flag in more regions incrementally by following the preview feature rollout SDP to manage risk.
+6. Do more testing and wait for a few weeks to ensure that the changes haven't introduced any regressions or unintended side effects.
+7. Create a new API version to expose the API changes and remove the feature flag.
+9. Update Swagger definition for the deployments and deployment operations APIs.
+10. Generate versions of Azure SDKs, including the Python SDK and C# SDK.
+11. Update client tools such as Azure CLI and Azure PowerShell to handle the new API responses.
+12. Collaborate with the Portal team to update the deployments blade, ensuring that the UI reflects the API changes.
+
+## Unresolved questions
+
+### Consistent schema V.S. Minimum backend changes
+
+To address the inconsistency in output resource schema between Azure resources and extensible resources, as highlighted in the Drawbacks section, a potential solution involves extending the schema to include additional properties for Azure resources as well, making Azure resources and extensible resources uniformly identifiable. For example:
+
+```diff
+{
+  ...
+  "type": "Microsoft.Resources/deployments",
+  "properties": {
+    "outputResources": [
+      // Azure resource
+      {
+        "id": "/subscriptions/00000000-0000-0000-0000-000000000001/resourceGroups/my-resource-group/providers/Microsoft.Storage/storageAccounts/my-storage-account"
++       "symbolicName": "myStorageAccount",
++       "resourceType": "Microsoft.Storage/storageAccounts",
++       "deploymentProvider": "az"
+      },
+      // Extensible resource
+      {
++       "symbolicName": "myService",
++       "resourceType": "core/Service@v1",
++       "deploymentProvider": "kubernetes"
+      }
+  }
+}
+```
+
+Implementing this enhancement involves comprehensive modifications across various deployment engine components, including but not limited to `AzureDeploymentEngine`, `ResourceProxyDefinition`, `DeploymentOutputResourceDefinition`, and numerous deployment jobs. Such widespread changes carry a significant risk due to the complexity of the deployment engine and the potential for unforeseen impacts on its functionality, which would require careful consideration and discussion.
+
+### Should the az provider always be returned?
+
+Building on the previous discussion, if the deploymentProvider details are not included within `outputResources` for Azure resources, should we consistently list the `az` provider in the `deploymentProviders` for sake of consistentcy and clarity?
+
+### Should `outputResources` include `existing` resources as well?
+
+Currently, `outputResources` only contains created or updated resources. Should we expand the property to include `existing` resources?
```

Comment:
To answer this question, I think it would be good to answer the question of "What do we expect people to do with the `outputResources` property?".

Reply:
We definitely have a usecase for this in Radius.

Reply:
This is one of the places where our fork of the DE deviated from the official one. We wanted to know which resources were *used* in a deployment and which ones were *mutated* by a deployment. 

This is information that Radius catalogs to help describe the application architecture.

Reply:
We discussed the issue today and decided to not include existing resources in `outputResources` since it will be a breaking change. To support Radius's use case, I think we can add a new property like `referencedResources`.